File: VB/CustomGrid.vb
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Drawing
Imports DevExpress.XtraGrid.Registrator
Imports DevExpress.XtraGrid.Skins
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Base.Handler
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.Handler
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Linq
Imports System.Windows.Forms

Namespace dxExample
	<ToolboxItem(True)>
	Public Class CustomGrid
		Inherits GridControl

		Protected Overrides Function CreateDefaultView() As BaseView
			Return CreateView("CustomGridView")
		End Function

		Protected Overrides Sub RegisterAvailableViewsCore(ByVal collection As InfoCollection)
			MyBase.RegisterAvailableViewsCore(collection)
			collection.Add(New CustomGridViewInfoRegistrator())
		End Sub
	End Class

	Public Class CustomGridViewInfoRegistrator
		Inherits GridInfoRegistrator

		Public Overrides ReadOnly Property ViewName() As String
			Get
				Return "CustomGridView"
			End Get
		End Property
		Private ps As CustomGridSkinPaintStyle
		Public Overrides Function PaintStyleByLookAndFeel(ByVal lookAndFeel As DevExpress.LookAndFeel.UserLookAndFeel, ByVal name As String) As ViewPaintStyle
			If ps Is Nothing Then
				ps = New CustomGridSkinPaintStyle()
				PaintStyles.Add(ps)
			End If
			Return ps
		End Function
		Public Overrides Function CreateView(ByVal grid As GridControl) As BaseView
			Return New CustomGridView(grid)
		End Function

		Public Overrides Function CreateHandler(ByVal view As BaseView) As BaseViewHandler
			Return New CustomGridViewHandler(TryCast(view, CustomGridView))
		End Function
	End Class
	Public Class CustomGridSkinPaintStyle
		Inherits GridSkinPaintStyle

		Public Sub New()
		End Sub
		Public Overrides Function CreateElementsPainter(ByVal view As BaseView) As DevExpress.XtraGrid.Views.Grid.Drawing.GridElementsPainter
			Return New CustomGridSkinElementsPainter(view)
		End Function
	End Class

	Public Class CustomGridSkinElementsPainter
		Inherits GridSkinElementsPainter

		Public Sub New(ByVal view As BaseView)
			MyBase.New(view)
		End Sub
		Protected Overrides Function CreateFilterPanelPainter() As GridFilterPanelPainter
			Dim painter As GridFilterPanelPainter = New CustomGridFilterPanelPainter(View)
			Return painter
		End Function
	End Class
	Public Class CustomGridFilterPanelPainter
		Inherits SkinGridFilterPanelPainter

		Public Sub New(ByVal provider As GridView)
			MyBase.New(provider)
			View = TryCast(provider, CustomGridView)
			customButtonPainter = New SkinEditorButtonPainter(DevExpress.LookAndFeel.UserLookAndFeel.Default.ActiveLookAndFeel)
		End Sub
		Protected Overrides Function CalcButtonLocation(ByVal client As Rectangle, ByVal size As Size, ByVal isRight As Boolean) As Point
			Dim point As Point = MyBase.CalcButtonLocation(client, size, isRight)
			If point.X < 10 Then
				point.X += View.UpdateButtonsRects() ' set the indent
			End If

			Return point
		End Function

		Public Overrides Sub DrawObject(ByVal e As ObjectInfoArgs)
			MyBase.DrawObject(e)
			DrawCustomButtons(e.Cache)
		End Sub

		Public Property View() As CustomGridView
		Private customButtonPainter As SkinEditorButtonPainter
		Private Sub DrawCustomButtons(ByVal cache As GraphicsCache)
			For Each button As EditorButton In View.ButtonsStrore.Keys
				customButtonPainter.DrawObject(GetButtonInfoArgs(cache, View.ButtonsStrore(button),button))
			Next button
		End Sub

		Friend Function GetButtonInfoArgs(ByVal cache As GraphicsCache, ByVal args As EditorButtonObjectInfoArgs, ByVal button As EditorButton) As EditorButtonObjectInfoArgs
			args.Cache = cache
			Dim state As ObjectState = ObjectState.Normal
			If TypeOf button.Tag Is ObjectState Then
				state = CType(button.Tag, ObjectState)
			End If
			args.State = state
			Return args
		End Function
	End Class

	Public Class CustomGridView
		Inherits GridView

		Public Sub New()
			ButtonsStrore = New Dictionary(Of EditorButton, EditorButtonObjectInfoArgs)()
		End Sub

		Public Sub New(ByVal grid As GridControl)
			MyBase.New(grid)
		End Sub
		Protected Overrides ReadOnly Property ViewName() As String
			Get
				Return "CustomGridView"
			End Get
		End Property
		Friend ButtonsStrore As Dictionary(Of EditorButton, EditorButtonObjectInfoArgs)

		Public Sub AddButton(ByVal b As EditorButton)
			ButtonsStrore.Add(b, New EditorButtonObjectInfoArgs(b, New DevExpress.Utils.AppearanceObject()))
		End Sub

		Friend Function UpdateButtonsRects() As Integer
			Dim offset As Integer = 10
			For Each button As EditorButton In ButtonsStrore.Keys
                Dim y As Integer = ViewInfo.ClientBounds.Y + (ViewInfo.ClientBounds.Height - ((ViewInfo.FilterPanel.Bounds.Height \ 8) * 7))
				Dim x As Integer = offset
				offset += 5 + button.Width
                Dim buttonRect As New Rectangle(x, y, button.Width, (ViewInfo.FilterPanel.Bounds.Height \ 8) * 6)
				ButtonsStrore(button).Bounds = buttonRect
			Next button
			Return offset
		End Function

		Public Event CustomButtonClick As EventHandler
		#Region "OnCustomButtonClick"
		''' <summary>
		''' Triggers the CustomButtonClick event.
		''' </summary>
		Public Overridable Sub OnCustomButtonClick(ByVal b As EditorButton, ByVal ea As EventArgs)
			Dim handler As EventHandler = CustomButtonClickEvent
			If handler IsNot Nothing Then
				handler(b, ea)
			End If
		End Sub
		#End Region

		Private shouldReset As Boolean = False
		Friend Sub UpdateButtonState(ByVal state As ObjectState, ByVal point As Point)
			Dim hitInfo As GridHitInfo = CalcHitInfo(point)
			If hitInfo.HitTest <> GridHitTest.FilterPanel Then
				If shouldReset Then
					shouldReset = False
					If ResetButtonState() Then
						InvalidateFilterPanel()
					End If
				End If
				Return
			End If
			UpdateButtonStateCore(state, point)
			InvalidateFilterPanel()
			shouldReset = True
		End Sub


		Private Function ResetButtonState() As Boolean
			Dim res As Boolean = False
			For Each button As EditorButton In ButtonsStrore.Keys
                If button.Tag IsNot Nothing AndAlso CType(button.Tag, ObjectState) <> ObjectState.Normal Then
                    button.Tag = ObjectState.Normal
                    res = True
                End If
			Next button
			Return res
		End Function

		Private Sub UpdateButtonStateCore(ByVal state As ObjectState, ByVal point As Point)
			For Each button As EditorButton In ButtonsStrore.Keys
				button.Tag = ObjectState.Normal
				If ButtonsStrore(button).Bounds.Contains(point) Then
					button.Tag = state
                    If state = ObjectState.Pressed Then
                        OnCustomButtonClick(button, EventArgs.Empty)
                    End If
				End If
			Next button
		End Sub
	End Class

	Public Class CustomGridViewHandler
		Inherits GridHandler

		Public Sub New(ByVal view As GridView)
			MyBase.New(view)
		End Sub
		Public ReadOnly Property View() As CustomGridView
			Get
				Return TryCast(MyBase.View, CustomGridView)
			End Get
		End Property
		Protected Overrides Function OnMouseDown(ByVal ev As MouseEventArgs) As Boolean
			View.UpdateButtonState(ObjectState.Pressed,ev.Location)
			Return MyBase.OnMouseDown(ev)
		End Function
		Protected Overrides Function OnMouseMove(ByVal ev As MouseEventArgs) As Boolean
			View.UpdateButtonState(ObjectState.Hot,ev.Location)
			Return MyBase.OnMouseMove(ev)
		End Function
		Protected Overrides Sub OnMouseLeave(ByVal e As EventArgs)
			View.UpdateButtonState(ObjectState.Normal, Point.Empty)
			MyBase.OnMouseLeave(e)
		End Sub
		Protected Overrides Function OnMouseUp(ByVal ev As MouseEventArgs) As Boolean
			View.UpdateButtonState(ObjectState.Normal,ev.Location)
			Return MyBase.OnMouseUp(ev)
		End Function
	End Class
End Namespace

File: CS/CustomGrid.cs
using DevExpress.Utils.Drawing;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Drawing;
using DevExpress.XtraGrid.Registrator;
using DevExpress.XtraGrid.Skins;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Base.Handler;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.Handler;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;

namespace dxExample
{
    [ToolboxItem(true)]
    public class CustomGrid : GridControl
    {
        protected override BaseView CreateDefaultView()
        {
            return CreateView("CustomGridView");
        }

        protected override void RegisterAvailableViewsCore(InfoCollection collection)
        {
            base.RegisterAvailableViewsCore(collection);
            collection.Add(new CustomGridViewInfoRegistrator());
        }
    }

    public class CustomGridViewInfoRegistrator : GridInfoRegistrator
    {
        public override string ViewName {   get{ return "CustomGridView"; }}
        CustomGridSkinPaintStyle ps;
        public override ViewPaintStyle PaintStyleByLookAndFeel(DevExpress.LookAndFeel.UserLookAndFeel lookAndFeel, string name)
        {
            if (ps == null)
            {
                ps = new CustomGridSkinPaintStyle();
                PaintStyles.Add(ps);
            }
            return ps;
        }
        public override BaseView CreateView(GridControl grid)
        {
            return new CustomGridView(grid);
        }

        public override BaseViewHandler CreateHandler(BaseView view)
        {
            return new CustomGridViewHandler(view as CustomGridView);
        }
    }
    public class CustomGridSkinPaintStyle : GridSkinPaintStyle
    {
        public CustomGridSkinPaintStyle()  { }
        public override DevExpress.XtraGrid.Views.Grid.Drawing.GridElementsPainter CreateElementsPainter(BaseView view)
        {
            return new CustomGridSkinElementsPainter(view);
        }
    }

    public class CustomGridSkinElementsPainter : GridSkinElementsPainter
    {
        public CustomGridSkinElementsPainter(BaseView view)  : base(view)   { }
        protected override ObjectPainter CreateFilterPanelPainter()
        {
            GridFilterPanelPainter painter = new CustomGridFilterPanelPainter(View);
            return painter;
        }
    }
    public class CustomGridFilterPanelPainter : SkinGridFilterPanelPainter
    {
        public CustomGridFilterPanelPainter(GridView provider): base(provider)
        {
            View = provider as CustomGridView;
            customButtonPainter = new SkinEditorButtonPainter(DevExpress.LookAndFeel.UserLookAndFeel.Default.ActiveLookAndFeel);
        }
        protected override Point CalcButtonLocation(Rectangle client, Size size, bool isRight)
        {
            Point point = base.CalcButtonLocation(client, size, isRight);
            if (point.X < 10)
                point.X += View.UpdateButtonsRects(); ; // set the indent
            return point;
        }

        public override void DrawObject(ObjectInfoArgs e)
        {
            base.DrawObject(e);
            DrawCustomButtons(e.Cache);
        }

        public CustomGridView View { get; set; }
        SkinEditorButtonPainter customButtonPainter;
        private void DrawCustomButtons(GraphicsCache cache)
        {
            foreach (EditorButton button in View.ButtonsStrore.Keys)
                customButtonPainter.DrawObject(GetButtonInfoArgs(cache, View.ButtonsStrore[button],button ));
        }

        internal EditorButtonObjectInfoArgs GetButtonInfoArgs(GraphicsCache cache, EditorButtonObjectInfoArgs args, EditorButton button)
        {
            args.Cache = cache;
            ObjectState state = ObjectState.Normal;
            if (button.Tag is ObjectState)
                state = (ObjectState)button.Tag;
            args.State = state;
            return args;
        }
    }

    public class CustomGridView : GridView
    {
        public CustomGridView()
        {
            ButtonsStrore = new Dictionary<EditorButton, EditorButtonObjectInfoArgs>();
        }

        public CustomGridView(GridControl grid): base(grid) { }
        protected override string ViewName{ get{return "CustomGridView"; } }
        internal Dictionary<EditorButton, EditorButtonObjectInfoArgs> ButtonsStrore;

        public void AddButton(EditorButton b)
        {
            ButtonsStrore.Add(b, new EditorButtonObjectInfoArgs(b, new DevExpress.Utils.AppearanceObject()));
        }

        internal int  UpdateButtonsRects()
        {
            int offset = 10;
            foreach (EditorButton button in ButtonsStrore.Keys)
            {
                int y = ViewInfo.ClientBounds.Y + (ViewInfo.ClientBounds.Height - (ViewInfo.FilterPanel.Bounds.Height / 8 * 7));
                int x = offset;
                offset += 5 + button.Width;
                Rectangle buttonRect = new Rectangle(x, y, button.Width, ViewInfo.FilterPanel.Bounds.Height / 8 *6);
                ButtonsStrore[button].Bounds = buttonRect;
            }
            return offset;
        }

        public event EventHandler CustomButtonClick;
        #region OnCustomButtonClick
        /// <summary>
        /// Triggers the CustomButtonClick event.
        /// </summary>
        public virtual void OnCustomButtonClick(EditorButton b, EventArgs ea)
        {
            EventHandler handler = CustomButtonClick;
            if (handler != null)
                handler(b, ea);
        }
        #endregion

        bool shouldReset = false;
        internal void UpdateButtonState(ObjectState state, Point point)
        {
            GridHitInfo hitInfo = CalcHitInfo(point);
            if (hitInfo.HitTest != GridHitTest.FilterPanel)
            {
                if (shouldReset)
                {
                    shouldReset = false;
                    if(ResetButtonState())
                        InvalidateFilterPanel(); 
                }
                return;
            }
            UpdateButtonStateCore(state, point);
            InvalidateFilterPanel();
            shouldReset = true;
        }


        private bool ResetButtonState()
        {
            bool res = false;
            foreach (EditorButton button in ButtonsStrore.Keys)
                if (button.Tag != null && (ObjectState)button.Tag != ObjectState.Normal)
                {
                    button.Tag = ObjectState.Normal;
                    res = true;
                }
            return res;
        }

        private void UpdateButtonStateCore(ObjectState state, Point point)
        {
            foreach (EditorButton button in ButtonsStrore.Keys)
            {
                button.Tag = ObjectState.Normal;
                if (ButtonsStrore[button].Bounds.Contains(point))
                {
                    button.Tag = state;
                    if (state == ObjectState.Pressed)
                        OnCustomButtonClick(button, EventArgs.Empty);
                }
            }
        }
    }

    public class CustomGridViewHandler : GridHandler
    {
        public CustomGridViewHandler(GridView view) : base(view) { }
        public CustomGridView View { get { return base.View as CustomGridView; } }
        protected override bool OnMouseDown(MouseEventArgs ev)
        {
            View.UpdateButtonState(ObjectState.Pressed,ev.Location);
            return base.OnMouseDown(ev);
        }
        protected override bool OnMouseMove(MouseEventArgs ev)
        {
            View.UpdateButtonState(ObjectState.Hot,ev.Location);
            return base.OnMouseMove(ev);
        }
        protected override void OnMouseLeave(EventArgs e)
        {
            View.UpdateButtonState(ObjectState.Normal, Point.Empty);
            base.OnMouseLeave(e);
        }
        protected override bool OnMouseUp(MouseEventArgs ev)
        {
            View.UpdateButtonState(ObjectState.Normal,ev.Location);
            return base.OnMouseUp(ev);
        }
    }
}
